<!-- TMPL_IF NAME=EXISTS --><!-- TMPL_IF NAME=COMPLETE --><!-- TMPL_INCLUDE NAME="header.html" -->

<script src="jquery-3.7.1.min.js"></script>
<script src="jquery.dataTables.min.js"></script>
<script src="dataTables.buttons.min.js"></script>
<script src="buttons.flash.min.js"></script>
<script src="buttons.html5.min.js"></script>
<script src="buttons.print.min.js"></script>
<script src="vfs_fonts.js"></script>
<script src="jszip.min.js"></script>


<h2 id="job_id">Results for job <!-- TMPL_VAR NAME=JOB_NAME ESCAPE=HTML --></h2>

<div id="resultstabs">

  <ul>
    <li id="tabletab" class="selected">Results Table</li>
    <li id="propertytab">Properties</li>
    <li id="biasestab">Biases</li>
  </ul>
  
</div>

<div class="resultdiv" id="tablediv">

<h3 id="hit_table_header">Hit table</h3>

<!-- TMPL_IF NAME=HIT_TABLE -->
<table class="results" id="results_table">
<!--<table class="results" id="results_table" style="width:100%"> -->
  <thead>
  <tr>
    <th>Gene Set</th>
    <th>Source</th>
    <th>Query count</th>
    <th>Background count</th>
    <th>Category size</th>
    <th>FDR</th>
    <th>Enrichment</th>
    <th>Potential bias</th>
  </tr>
  </thead>
  <tbody>
  <!-- TMPL_LOOP NAME=HIT_TABLE -->
  <tr>
    <td><!-- TMPL_VAR NAME=GO_NAME ESCAPE=HTML --></td>
    <td><!-- TMPL_VAR NAME=SOURCE ESCAPE=HTML --></td>
    <td><!-- TMPL_VAR NAME=QUERY_COUNT ESCAPE=HTML --></td>
    <td><!-- TMPL_VAR NAME=BACKGROUND_COUNT ESCAPE=HTML --></td>
    <td><!-- TMPL_VAR NAME=CATEGORY_COUNT ESCAPE=HTML --></td>
    <td><!-- TMPL_VAR NAME=FDR ESCAPE=HTML --></td>
    <td><!-- TMPL_VAR NAME=ENRICHMENT ESCAPE=HTML --></td>
    <td>
      <!-- TMPL_LOOP NAME=POTENTIAL_BIAS -->
      <span id="potential_bias" class="bias <!-- TMPL_VAR NAME=CLASS ESCAPE=HTML -->"><!-- TMPL_VAR NAME=BIAS ESCAPE=HTML --></span>
      <!-- /TMPL_LOOP -->
    </td>
  </tr> 
  <!-- /TMPL_LOOP -->
  </tbody>
  <tfoot>
    <tr>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
    <tr>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </tfoot> 
</table>
<!-- TMPL_ELSE -->
<p class=center>No significant category hits</p>

<!-- /TMPL_IF -->

</div>

<!-- The content for the search properties -->
<div class="resultdiv" id="propertydiv">
<h3>Query and Background Properties</h3>
<br><br>

<table class="prop" id="query_properties" name="properties_info">
  <!-- TMPL_LOOP NAME=PROPERTIES_INFO -->
  <tr>
    <th>Species</th>
    <td><!-- TMPL_VAR NAME=SPECIES ESCAPE=HTML --></td>
  </tr>
  <tr>
    <th>Minimum number of genes in functional category</th>
    <td><!-- TMPL_VAR NAME=MIN_CATEGORY_SIZE ESCAPE=HTML --></td>
  </tr>
  <tr>  
    <th>Maximum number of genes in functional category</th>
    <td><!-- TMPL_VAR NAME=MAX_CATEGORY_SIZE ESCAPE=HTML --></td>
  </tr>
  <tr>  
    <th>Number of query genes entered</th>
    <td><!-- TMPL_VAR NAME=QUERY_NO ESCAPE=HTML --></td>
  </tr>
  <tr>  
    <th>Number of background genes entered</th>
    <td><!-- TMPL_VAR NAME=BG_NO ESCAPE=HTML --></td>
  </tr>
  <tr>  
    <th>Job id</th>
    <td><!-- TMPL_VAR NAME=JOB ESCAPE=HTML --></td>
  </tr>
  <!-- /TMPL_LOOP -->
</table>
<br>
<br>
<h4 id="properties_text">
The plots below show properties of the set of query genes that were entered compared
to the genes used as the background set.
</h4>

<br>
<p class="graph center">
  <img src="<!-- TMPL_VAR NAME=GENELENGTH_GRAPH ESCAPE=HTML -->" alt="Gene Length Graph">
</p>
<br>
<p class="graph center">
  <img src="<!-- TMPL_VAR NAME=GC_GRAPH ESCAPE=HTML -->" alt="GC Content Graph">
</p>
<br>
<p class="graph center">
  <img src="<!-- TMPL_VAR NAME=CHROMOSOME_GRAPH ESCAPE=HTML -->" alt="Chromosome Graph">
</p>
<br>
<p id="gene_set_source_text">
The properties of the genes were derived from <a href="https://www.ensembl.org/info/data/ftp/index.html" target = "_blank"> Ensembl gtf files</a>
<br>
Gene set files used in the enrichment analysis were from the <a href="http://download.baderlab.org/EM_Genesets/" target = "_blank"> Bader Lab</a>
</p>
<br>
</div>


<!-- The content for the biases -->
<div class="resultdiv" id="biasesdiv">
<h3>Biases</h3>
<!-- TMPL_IF NAME=HIT_TABLE -->
<table class="biasSummary" id="biasSummaryTable">
  <tr>
    <th>Bias Source</th>
    <th>Number flagged</th>
    <th>Total in bias category</th>
  </tr>
  <!-- TMPL_LOOP NAME=BIAS_SUMMARY -->
  <tr>
    <td><!-- TMPL_VAR NAME=BIAS_SOURCE ESCAPE=HTML --></td>
    <td><!-- TMPL_VAR NAME=TOTAL_FLAGGED ESCAPE=HTML --></td>
    <td><!-- TMPL_VAR NAME=TOTAL_IN_CATEGORY ESCAPE=HTML --></td>
  </tr>
  <!-- /TMPL_LOOP -->


</table>
<!-- TMPL_ELSE -->
<p class=center>No significant category hits</p>

<!-- /TMPL_IF -->



<p class=left id="documentation_links">
  These documents describe how the lists of biased genes were produced for mouse.
  <br>
  <br>
  <a href="http://htmlpreview.github.io/?https://github.com/laurabiggins/GOcategoryStats/blob/master/doc/1.length_biased.html" target="_blank">Length biased genes</a>
  <br>
  <br>
  <a href="https://raw.githack.com/laurabiggins/GOcategoryStats/master/doc/2.high_transcript_biased.html" target="_blank">Genes with high numbers of transcripts</a>
  <br>
  <br>
  <a href="https://raw.githack.com/laurabiggins/GOcategoryStats/master/doc/3.GC_content_biased.html" target="_blank">GC content biased</a>
  <br>
  <br>
  <a href="http://htmlpreview.github.io/?https://github.com/laurabiggins/GOcategoryStats/blob/master/doc/4.chromosome_biased.html" target="_blank">Chromosome biased</a>
  <br>
  <br>
  <a href="http://htmlpreview.github.io/?https://github.com/laurabiggins/GOcategoryStats/blob/master/doc/5.closest_gene.html" target="_blank">Closest gene to random positions</a>
<!--  <br> 
  <br>
  <a href="http://htmlpreview.github.io/?https://github.com/laurabiggins/GOcategoryStats/blob/master/doc/6.1_public_data_tidy.html" target="_blank">Public data tidy</a>
  <br>
  <br>  
  <a href="http://htmlpreview.github.io/?https://github.com/laurabiggins/GOcategoryStats/blob/master/doc/7.GO_category_summary.html" target="_blank">GO category summary</a>
  <br>
  <br>  -->
</p>
</div>

<!-- The biasexplanation div which will pop up if you click on a bias -->

<div id="biasexplanation">

</div>

<!-- TMPL_ELSE --><!-- TMPL_INCLUDE NAME="refresh_header.html" -->

<h2>Job <!-- TMPL_VAR NAME=JOB_ID ESCAPE=HTML --> is running...</h2>

<p class=center>
  <img src="goingliath_logo.gif">
</p>

<p class=center>
  This page will refresh in 10 seconds, or you can <a href="javascript:history.go(0)">check again now</a>.
</p>

<!-- /TMPL_IF --><!-- TMPL_ELSE --><!-- TMPL_INCLUDE NAME="header.html" -->

<h2>Job <!-- TMPL_VAR NAME=JOB_ID ESCAPE=HTML --> doesn't exist</h2>

<p class="center">
  GOliath jobs are only kept on our server for around 24hours after they are run.  After this time you'll have to run them again to see the results.
</p>

<p class="center">
Sorry about that.
</p>

<!-- /TMPL_IF --> 


<!-- TMPL_INCLUDE NAME="javascript.html" -->
<!-- TMPL_INCLUDE NAME="footer.html" -->

